File: lib/ime/ime_edit.dart
/// Pure editing rules shared with the Kotlin IME (keep behavior aligned).
class ImeEdit {
  static bool composingInvalid({
    required int newSelStart,
    required int newSelEnd,
    required int candStart,
    required int candEnd,
  }) {
    if (candStart < 0 || candEnd < 0) return true;
    if (newSelStart != newSelEnd) return true;
    return newSelStart < candStart || newSelStart > candEnd;
  }

  static int clusterLen(String text) {
    if (text.isEmpty) return 0;
    var i = text.length;
    while (i > 0 && _isMark(text.codeUnitAt(i - 1))) {
      i--;
    }
    if (i == 0) return text.length;
    i--;
    while (i >= 2 &&
        text.codeUnitAt(i - 1) == 0x09CD &&
        _isBnCons(text.codeUnitAt(i - 2))) {
      i -= 2;
    }
    return text.length - i;
  }

  static bool _isMark(int c) =>
      (c >= 0x0981 && c <= 0x0983) ||
      c == 0x09BC ||
      c == 0x09BE ||
      (c >= 0x09BF && c <= 0x09C4) ||
      (c >= 0x09C7 && c <= 0x09C8) ||
      (c >= 0x09CB && c <= 0x09CD) ||
      c == 0x09D7;

  static bool _isBnCons(int c) =>
      (c >= 0x0995 && c <= 0x09B9) ||
      (c >= 0x09DC && c <= 0x09DF) ||
      c == 0x09CE;
}

/// In-memory Avro IME session used by tests.
class ImeSession {
  ImeSession(this.parse);

  final String Function(String) parse;
  String buffer = '';
  String committed = '';
  String composing = '';
  String mode = 'avro';

  void type(String ch) {
    if (mode != 'avro') {
      committed += ch;
      return;
    }
    if (ch == ' ' || (ch.length == 1 && !_isLetter(ch) && ch != '`')) {
      committed += parse(buffer) + ch;
      buffer = '';
      composing = '';
      return;
    }
    buffer += ch;
    composing = parse(buffer);
  }

  void backspace() {
    if (buffer.isNotEmpty) {
      buffer = buffer.substring(0, buffer.length - 1);
      composing = buffer.isEmpty ? '' : parse(buffer);
      return;
    }
    if (committed.isEmpty) return;
    final n = ImeEdit.clusterLen(committed);
    committed = committed.substring(0, committed.length - n);
  }

  void enter() {
    committed += parse(buffer);
    if (mode == 'avro' && buffer.isNotEmpty) {
      // word finalized; newline is a boundary
    }
    buffer = '';
    composing = '';
    committed += '\n';
  }

  void switchMode(String next) {
    committed += parse(buffer);
    buffer = '';
    composing = '';
    mode = next;
  }

  void fieldChanged() {
    buffer = '';
    composing = '';
  }

  bool get isComposing => buffer.isNotEmpty;

  static bool _isLetter(String ch) {
    final c = ch.codeUnitAt(0);
    return (c >= 65 && c <= 90) || (c >= 97 && c <= 122);
  }
}

File: android/app/src/main/kotlin/com/avrokeyboard/app/ime/ImeEdit.kt
package com.avrokeyboard.app.ime

import android.text.InputType
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection

object ImeEdit {
    fun isPassword(info: EditorInfo?): Boolean {
        if (info == null) return false
        val variation = info.inputType and InputType.TYPE_MASK_VARIATION
        val klass = info.inputType and InputType.TYPE_MASK_CLASS
        if (klass == InputType.TYPE_CLASS_TEXT) {
            return variation == InputType.TYPE_TEXT_VARIATION_PASSWORD ||
                variation == InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD ||
                variation == InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
        }
        if (klass == InputType.TYPE_CLASS_NUMBER) {
            return variation == InputType.TYPE_NUMBER_VARIATION_PASSWORD
        }
        return false
    }

    fun isMultiline(info: EditorInfo?): Boolean {
        if (info == null) return false
        return info.inputType and InputType.TYPE_TEXT_FLAG_MULTI_LINE != 0
    }

    fun enter(ic: InputConnection, info: EditorInfo?) {
        val options = info?.imeOptions ?: EditorInfo.IME_ACTION_UNSPECIFIED
        val action = options and EditorInfo.IME_MASK_ACTION
        val noEnterAction = options and EditorInfo.IME_FLAG_NO_ENTER_ACTION != 0
        if (isMultiline(info) || noEnterAction) {
            ic.commitText("\n", 1)
            return
        }
        when (action) {
            EditorInfo.IME_ACTION_NONE, EditorInfo.IME_ACTION_UNSPECIFIED ->
                ic.commitText("\n", 1)
            else -> ic.performEditorAction(action)
        }
    }

    fun deleteBefore(ic: InputConnection) {
        val selected = ic.getSelectedText(0)
        if (!selected.isNullOrEmpty()) {
            ic.commitText("", 1)
            return
        }
        val before = ic.getTextBeforeCursor(12, 0)?.toString().orEmpty()
        if (before.isEmpty()) {
            ic.deleteSurroundingText(1, 0)
            return
        }
        ic.deleteSurroundingText(clusterLen(before), 0)
    }

    fun clusterLen(text: String): Int {
        if (text.isEmpty()) return 0
        var i = text.length
        while (i > 0 && isMark(text[i - 1])) i--
        if (i == 0) return text.length
        i--
        while (i >= 2 && text[i - 1] == HASANTA && isBnCons(text[i - 2])) {
            i -= 2
        }
        return text.length - i
    }

    fun composingInvalid(newSelStart: Int, newSelEnd: Int, candStart: Int, candEnd: Int): Boolean {
        if (candStart < 0 || candEnd < 0) return true
        if (newSelStart != newSelEnd) return true
        return newSelStart < candStart || newSelStart > candEnd
    }

    private const val HASANTA = '\u09CD'

    private fun isMark(ch: Char): Boolean {
        val c = ch.code
        return c in 0x0981..0x0983 ||
            c == 0x09BC ||
            c == 0x09BE ||
            c in 0x09BF..0x09C4 ||
            c in 0x09C7..0x09C8 ||
            c in 0x09CB..0x09CD ||
            c == 0x09D7
    }

    private fun isBnCons(ch: Char): Boolean {
        val c = ch.code
        return c in 0x0995..0x09B9 || c in 0x09DC..0x09DF || c == 0x09CE
    }
}
